### ISSUE TITLE
keras seq2seq example

### ISSUE THREAD

**Issue by alwaysprep** [keras-team/keras#8338]
Hi,

If I remember correctly, using TimeDistributed and Dense layer after return_sequences was a good practice. Is there a reason to use just dense layer here, or am I missing something?

https://github.com/fchollet/keras/blob/09e0044882f9311d5ea227e130209636defabfa7/examples/lstm_seq2seq.py#L138

Thanks,

**Comment by litpuvn** [comment]
Hi,
I am not sure why in the inference mode, weights of saved seq2seq model are not loaded to ``encoder_model`` and ``decoder_model``.
is there any reason for it?
thanks

**Comment by talorwu** [comment]
yes...I am also puzzled with it. And I change it to : 
`decoder_dense = TimeDistributed(Dense(num_decoder_tokens, activation='softmax'))`
and it also work , and I think just use Dense on time seq is not correct, because the input_shape is not correct. But it is work, so confused

**Comment by kaushalshetty** [comment]
@alwaysprep @talorwu The  decoder_outputs from decoder_lstm layer  has shape 
(None,None,256).Since there is no timestep involved in this tensor a single Dense layer would suffice instead of a Timedistributed layer. Timedistributed layer would be required if the tensor was of shape
(None,timestep,dim). Try analyzing `model.summary(),encoder_model.summary(),decoder_model.summary() `to get a clearer picture on the playing tensors.

@litpuvn the model is earlier compiled and fit before making use of it during inference. Basically during training the model weights are updated. Then we make the inference models using the layers of the earlier fitted model which already has trained weights. 


Thanks

**Comment by talorwu** [comment]
So...If I use TimeDistributed, there will be less variables , because TimeDistributed will share the variables on each timesteps. it's right?

**Comment by kaushalshetty** [comment]
Yes the same weights are shared in each timestep as per theory.

**Comment by kaushalshetty** [comment]
@talorwu I was testing this piece of code:
```
from keras.models import Model
from keras.layers import Input,Dense,LSTM,SimpleRNN,TimeDistributed
input_test = Input(shape=(21,100))
lstm_test = LSTM(output_dim = 30,input_shape=(21,100),return_sequences=True)(input_test)

dense_output =TimeDistributed(Dense(100,activation='softmax'))(lstm_test)


model = Model(input_test,dense_output)
model.summary()
```

and 

```
from keras.models import Model
from keras.layers import Input,Dense,LSTM,SimpleRNN,TimeDistributed
input_test = Input(shape=(21,100))
lstm_test = LSTM(output_dim = 30,input_shape=(21,100),return_sequences=True)(input_test)

dense_output =Dense(100,activation='softmax')(lstm_test)


model = Model(input_test,dense_output)
model.summary()

```

I could see that code1 and code2 ran successfully in keras versions 1.2.2
but code2  failed in 1.2.1 version(ERROR: Dense expecting ndim=2 not 3). 
Therefore Dense can replace TimeDistributed(Dense()) is what I think. Please  correct me if I am wrong.

**Comment by talorwu** [comment]
@kaushalshetty emmm....I think you're ALMOST right. But I think these two ways is not as like as two peas. Just like you mentioned above  the same weights will be shared in each timestep when you use TimeDistributed Wrapper. And I think it seems `Flatten()` firstly and then `Dense` if use Dense directly

**Comment by talorwu** [comment]
![39114243048398181](https://user-images.githubusercontent.com/7067195/33267363-27c25fcc-d3b4-11e7-911d-c76c776a7f35.jpg)

**Comment by GarryChien** [comment]
@kaushalshetty  You mentioned the model is earlier compiled and fit before making use of it during inference. That's why the encoder model can be initialized in such fashion. What if we want to reload the trained model from another script after training. How should we effectively and properly setup the models for inference then? Thanks in advance!
